localstack/aws/protocol/serializer.py

```diff
@@ -153,15 +159,38 @@ def serialize_error_to_response(
             status_code = error.status_code
             shape = None
         else:
-            # It it's not a CommonServiceException, the exception is being serialized based on the specification
+            # It's not a CommonServiceException, the exception is being serialized based on the specification
 
             # The shape name is equal to the class name (since the classes are generated from the shape's name)
             error_shape_name = error.__class__.__name__
 
-            # Lookup the corresponding error shape in the operation model
-            shape = next(
-                shape for shape in operation_model.error_shapes if shape.name == error_shape_name
-            )
+            try:
+                # Lookup the corresponding error shape in the operation model
+                shape = next(
+                    shape
+                    for shape in operation_model.error_shapes
+                    if shape.name == error_shape_name
+                )
+            except StopIteration:
+                LOG.warning(
+                    "Error shape %s not found for operation %s. Falling back to search across operations.",
+                    error_shape_name,
+                    operation_model.name,
+                )
+                try:
+                    # The exception to serialize is not defined for the specific operation.
+                    # Look for the error shape across all operations in the service.
+                    shape = next(
+                        shape
+                        for shape in operation_model.service_model.error_shapes
+                        if shape.name == error_shape_name
+                    )
```

Comment:
could we maybe simplify this entire method by just calling `operation_model.service_model.shape_for_error_code("..")` ? is there any real benefit of going through the exercise of checking the operation errors first if we're doing a fallback to the service anyway? the service model should contain all error shapes.

`service.shape_for_error_code("NoSuchUpload")` will return the error shape, no need to do the iteration.

or am i missing something?

Reply:
That's a really good catch. Actually, the `shape_for_error_code` is not exactly doing the same thing (since the exception's name is the shape's name not the error code, they can differ sometimes). But I switched to `service.shape_for()` instead.